src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -89,25 +89,28 @@ private static void ReadCollectionItem(ODataReader reader, Stack<ODataItemBase>
                     }
                     else
                     {
-                        ODataItemBase parentItem = itemsStack.Peek();
-                        ODataResourceSetWrapper parentResourceSet = parentItem as ODataResourceSetWrapper;
-                        if (parentResourceSet != null)
-                        {
-                            parentResourceSet.Resources.Add(resourceWrapper);
-                        }
-                        else
-                        {
-                            ODataNestedResourceInfoWrapper parentNestedResource = (ODataNestedResourceInfoWrapper)parentItem;
-                            Contract.Assert(parentNestedResource.NestedResourceInfo.IsCollection == false, "Only singleton nested properties can contain resource as their child.");
-                            Contract.Assert(parentNestedResource.NestedItems.Count == 0, "Each nested property can contain only one resource as its direct child.");
-                            parentNestedResource.NestedItems.Add(resourceWrapper);
-                        }
+                        AddResourceToParent(itemsStack, resourceWrapper);
                     }
 
                     itemsStack.Push(resourceWrapper);
                     break;
 
+                case ODataReaderState.DeletedResourceStart:
+                    ODataDeletedResource deletedResource = (ODataDeletedResource)reader.Item;
+                    Contract.Assert(deletedResource != null, "Deleted resource should not be null");
+
+                    ODataResourceWrapper deletedResourceWrapper = new ODataResourceWrapper(deletedResource);
+
+                    Contract.Assert(itemsStack.Count != 0, "Deleted Resource should not be top level item");
```

Comment:
If we're checking that there are items in the stack => `itemsStack.Count != 0`, is the accompanying assertion message valid => `"Deleted Resource should not be top level item"`?
Or should we be saying something like `"Deleted resource should not be the bottom level item"` or `"Deleted resource should be the top level item"`?

Reply:
Here we are giving a reason why the `itemstack.count != 0`. The reason is because, the `deleted resource should not be a top level item`. The accompanying message is correct.